src/stash/src/postgres.rs

```diff
@@ -61,24 +64,38 @@ CREATE TABLE uppers (
 /// tables are not specified and should not be relied upon. The only promise is
 /// stability. Any changes to the table schemas will be accompanied by a clear
 pub struct Postgres {
-    conn: Client,
+    url: String,
+    client: Option<Client>,
     epoch: i64,
+    rng: [u8; 16],
 }
 
 impl std::fmt::Debug for Postgres {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         f.debug_struct("Postgres")
+            .field("url", &self.url)
             .field("epoch", &self.epoch)
+            .field("rng", &self.rng)
             .finish_non_exhaustive()
     }
 }
 
 impl Postgres {
     /// Opens the stash stored at the specified path.
-    pub async fn open(mut conn: Client) -> Result<Postgres, StashError> {
-        conn.batch_execute("SET default_transaction_isolation = serializable")
-            .await?;
-        let tx = conn.transaction().await?;
+    pub async fn new(url: String) -> Result<Postgres, StashError> {
+        let mut conn = Postgres {
+            url,
+            client: None,
+            epoch: 0,
+            // The call to rand::random here assumes that the rng's seed source is from a
+            // secure source that will differ per thread. The docs for ThreadRng say it "is
```

Comment:
does it need to be secure? or just sufficiently random? I'll admit, I am not 100% sure what this rng does that's different than the epoch

Reply:
This is explained below: if the database is dropped and recreated by another connection then it could share the same epoch, causing the connection here to think it's still connected to a datastore for which it assumes exclusive access. The rng bytes add an additional check such that even if another connection generates the same epoch, it won't generate the same rng.

Reply:
This is only expected to be a problem in tests, right? Like, if the schema gets dropped and recreated in production, we've just wiped out our customer's entire installation, right?

Reply:
Yes, that sounds correct. But it still felt worthwhile building in for like, uh, antagonistic jepsen testing or something. I tried to implement it such that it's not any additional queries in the `transact` code path: it's just an additional column that can be quickly checked.